### ISSUE TITLE
Smallest-first ordering governs both admission and eviction, worth a look given the SkyRL/Dynamo rollout integration

### ISSUE THREAD

**Issue by SARPAT** [ThunderAgent-org/ThunderAgent#46]
Hi, I've been running ThunderAgent as an external user and measuring its
scheduling behaviour. I found two related things worth flagging together,
since they come from the same root mechanism.

**The short version:** eviction is smallest-first (correct, proven optimal by
Lemma 4.1). Admission of brand-new programs turns out to use the same
ordering. Under burst arrival, that produces very large head-of-line blocking
for large-prompt programs, and it interacts specifically badly with
group-structured RL rollouts.

### 1. Admission ordering for new programs

Under burst arrival (64 programs launched over ~13s against one backend),
time-to-first-response scales almost linearly with prompt size:

| Start tokens | Time to first response |
|---|---|
| 1,444 | **0.5 s** |
| 3,477 | 103.9 s |
| 5,186 | 142.4 s |
| 7,092 | 303.0 s |
| 8,046 | **404.4 s** |

corr(start_tokens, first_response_latency) = +0.855 / +0.745 / +0.870 across three seeds ([data](https://github.com/SARPAT/thunderagent-churn-study/blob/57ba070ae6e794931c5451c8ccf5cf16100f964b/results/grpo_run.json), [script](https://github.com/SARPAT/thunderagent-churn-study/blob/57ba070ae6e794931c5451c8ccf5cf16100f964b/analysis/admission_check.py#L22-L31)). About 75% of programs waited over 20s to start ([48/64, 51/64, 47/64](https://github.com/SARPAT/thunderagent-churn-study/blob/57ba070ae6e794931c5451c8ccf5cf16100f964b/analysis/step0_check.py#L23-L27)). Every subsequent step averaged 0.6s, so this is purely a first-request effect.

**Source.** In
[`_resume_paused_programs`, line 846](https://github.com/ThunderAgent-org/ThunderAgent/blob/7ddc8610270e56d3b109eed8796b3a4360fc67c9/ThunderAgent/scheduler/router.py#L846),
all waiting programs are sorted globally before priority grouping:

```python
paused_programs = self._get_paused_programs_sorted(ascending=True)
```
'The Dynamo port's own docs describe the same behavior: ["When utilization drops it resumes the smallest-token programs first, BFD-packing them back under threshold."](https://docs.nvidia.com/dynamo/dev/user-guides/agents/thunder-agent-program-scheduler)

`new_program_group` inherits that ordering, so brand-new programs are admitted
smallest-first, same as the eviction path. Under a burst, large-prompt
programs land at the back of the queue and stay there.

### 2. The scheduler has no concept of an RL group

In GRPO-style training, a group of rollouts shares one prompt, and advantages
cannot be computed until all members finish. I could not find any group,
batch, or trajectory-set concept anywhere in the scheduler, only status
buckets (reasoning / new / acting). Because group members share a prompt, they
also share similar token counts, so the smallest-first rule does not spread
within a group, it targets **whole groups**.

`corr(group size, evictions on that group)` across three seeds:
**-0.840, -0.841, -0.900**
([analysis](https://github.com/SARPAT/thunderagent-churn-study/blob/57ba070ae6e794931c5451c8ccf5cf16100f964b/analysis/grpo_analysis.py)).

### Both effects together, on one pair of groups

Same run, same 43,216-token capacity:

| | Small-prompt group | Large-prompt group |
|---|---|---|
| Prompt size | 1,504 tok | 8,120 tok |
| Time to admit | **2.0 s** | **447.2 s** |
| Time to execute | 102.6 s | 71.4 s |
| **Total** | 104.6 s | **518.5 s** |

([data](https://github.com/SARPAT/thunderagent-churn-study/blob/57ba070ae6e794931c5451c8ccf5cf16100f964b/results/grpo_run.json), [phase-split analysis](https://github.com/SARPAT/thunderagent-churn-study/blob/57ba070ae6e794931c5451c8ccf5cf16100f964b/analysis/separate_effects.py#L76-L85))

The small group gets in instantly and pays on the back end (18 evictions in this run). The large group is never evicted once, but pays almost entirely on the front end. Checked these aren't just two views of the same measurement: [partial correlation](https://github.com/SARPAT/thunderagent-churn-study/blob/57ba070ae6e794931c5451c8ccf5cf16100f964b/analysis/why_faster.py#L96-L103) confirms admission wait and execution-phase eviction cost are statistically independent, each driven by a different variable (size drives the former, eviction count drives the latter).

### Why I think this is worth your attention specifically

Burst arrival is not a synthetic artifact, it's the normal shape of RL rollout
launch: every trajectory in a batch fires at the start of a training step.
ThunderAgent is already the rollout gateway in SkyRL ([PR #1645](https://github.com/NovaSky-AI/SkyRL/pull/1645?utm_source=chatgpt.com), 3x on 40 H100
GPUs) and integrated into NVIDIA Dynamo (PR [#9448](https://github.com/ai-dynamo/dynamo/pull/9448?utm_source=chatgpt.com), program abstraction as a
first-class scheduling unit, PR [#8789](https://github.com/ai-dynamo/dynamo/pull/8789?utm_source=chatgpt.com)). In that setting, whole GRPO groups
built from short prompts can end up disproportionately delayed at admission,
not eviction, which is a different mechanism than the one the paper's
eviction analysis covers.

<h3>Tested a change: FIFO admission</h3>

<p>Since the mechanism is a single sort call, I patched just <code>new_program_group</code> behind a flag (leaving resume ordering for already-running programs untouched) and compared three orderings across three seeds each, same workload, same eviction path:</p>

<table>
<thead>
<tr><th></th><th>Smallest-first (upstream)</th><th>FIFO</th></tr>
</thead>
<tbody>
<tr><td>corr(size, TTFT)</td><td>+0.855 / +0.747 / +0.868</td><td>+0.272 / +0.230 / +0.406</td></tr>
<tr><td>Large/small wait ratio</td><td>9.4x / 8.7x / 9.8x</td><td>4.8x / 3.9x / 2.6x</td></tr>
<tr><td>Makespan</td><td>526 / 512 / 516 s</td><td><strong>504 / 495 / 498 s</strong></td></tr>
<tr><td>Worst group completion</td><td>518 / 496 / 503 s</td><td><strong>488 / 474 / 480 s</strong></td></tr>
</tbody>
</table>

<p>FIFO won on makespan and worst-group completion in every seed, while cutting the size-fairness ratio roughly in half. It costs about 20 to 30 seconds of mean group completion, and total eviction count rose somewhat (checked whether this concentrated on any one program; it didn't, the extra events were spread rather than piled up).</p>

<p>The pattern lines up with standard scheduling theory: smallest-first minimizes mean flow time, which is what upstream actually optimizes for. Longest-processing-time orderings minimize makespan. FIFO landed as a strong middle ground, and empirically the best of what I tested on the metric that matters for synchronous rollouts, the slowest group in the batch, not the average one.</p>

<p>A third arm, largest-first, looked even faster on makespan in one seed (-7.1%) but that did not hold up in the other two (+0.6%, -1.6%), so I'd treat that as noise rather than a real effect.</p>

### Scope and caveats

- 1x NVIDIA L4 (23GB), vLLM 0.27.1, Qwen3-0.6B, single backend
- KV capacity 43,216 tokens, contexts 1,500 to 8,000 tokens, 3 seeds
- Far smaller than your 8x H100 evaluations. The ordering mechanism is
  confirmed in source, so it should hold at any scale, but the absolute wait
  times here should not be read as production numbers.

Full setup, logs, and every script:
**https://github.com/SARPAT/thunderagent-churn-study**

### Questions

1. Is smallest-first admission for brand-new programs deliberate, or a side
   effect of `new_program_group` reusing the resume path's global sort?
2. Given the SkyRL/Dynamo integrations, has group-completion time (rather than
   per-request throughput) come up as a scheduling objective internally? Happy
   to share the full GRPO-shaped workload setup if useful.

---

**Comment by HaoKang-Timmy** [comment]
Hi, thanks for paying attention to our project
Group order scheduler and reusing is a proeject that we are focusing on, will be released soon!
Thanks for providing the script. I will respond to you with my earilest convenience.
